### PR TITLE
fix: capture complete extension load time

### PR DESCRIPTION
**Pull request by peternhale**
- Add utility function to vscode-utils that returns the activation event for a given extension from extension host log
- At the beginning of each extension’s activate function
-- Capture the timestamp of function entry
-- Capture the timestamp of the extension activation event using util from vscode-utils
- Send activation telemetry record with
-- Time spent in activation function (this is the current data)
-- Timestamp of activate function entry
-- Timestamp of extension activation
-- Time spent for activation (activate function exit timestamp - extension activation event timestamp) in milliseconds

AC:

- vscode-utils provides a function that returns the ext host extension activation timestamp for the provided extension id
- sendActivation telemetry method will accept
-- Time spent in activation function (this is the current data)
-- Timestamp of activate function entry
-- Timestamp of extension activation
-- Time spent for activation (activate function exit timestamp - extension activation event timestamp) in milliseconds
- One should be able to visit AppInsights to verify the new data are available in the activation telemetry record

@W-14915912@

### CODE REVIEW

File: packages/salesforcedx-vscode-lightning/src/telemetry/index.ts [hunk 1]
```diff
@@ -0,0 +1,8 @@
+/*
+ * Copyright (c) 2024, salesforce.com, inc.
+ * All rights reserved.
+ * Licensed under the BSD 3-Clause license.
+ * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
+ */
+import { TelemetryService } from '@salesforce/salesforcedx-utils-vscode';
+export const telemetryService = TelemetryService.getInstance('salesforcedx-vscode-apex');
```
Comment (mingxuanzhangsfdx) [inline packages/salesforcedx-vscode-lightning/src/telemetry/index.ts:8]: Is this the right name for lightning extension?

File: packages/salesforcedx-vscode-apex/src/telemetry/index.ts [hunk 2]
```diff
@@ -5,4 +5,4 @@
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 import { TelemetryService } from '@salesforce/salesforcedx-utils-vscode';
-export const telemetryService = TelemetryService.getInstance();
+export const telemetryService = TelemetryService.getInstance('salesforcedx-vscode-apex');
```
Comment (mingxuanzhangsfdx) [inline packages/salesforcedx-vscode-apex/src/telemetry/index.ts:8]: I do not feel it is necessary to specify a name for the core extensions such as apex & lightning. The reason why we pass in the name is only that e4d is using the telemetryService by salesforcedx-vscode-core. However, extensions in this core repo use their own TelemetryService 
`import { TelemetryService } from '@salesforce/salesforcedx-utils-vscode';` makes it.

Comment (mingxuanzhangsfdx) [reply packages/salesforcedx-vscode-apex/src/telemetry/index.ts:8]: Of course, it does not hurt to use in this way. Actually the name is specified during initializeService for now (not ideally).

**Review by mingxuanzhangsfdx — CHANGES_REQUESTED** [review]
It does not work in apex debugger. It is also missed in visualforce & replay debugger.

File: packages/salesforcedx-utils-vscode/src/helpers/activationTracker.ts [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+    this._activationInfo = {
+      ...this._activationInfo,
+      ...extensionInfo,
+      activationTime
+    };
+    this.telemetryService.sendActivationEventInfo(this.activationInfo);
+  }
+
+  public get activationInfo(): ActivationInfo {
```
[the remark this answers is not in the document]
Comment (peternhale) [reply packages/salesforcedx-utils-vscode/src/helpers/activationTracker.ts:63]: This pattern prevents direct property access and forces it through logic.

If we wanted to allow a property's value to change we would add
```
public set activationInfo(activationInfo: ActivationInfo): void {
  // do some validation checks on the value to be stored. Allows the class to control how changes are made
  if (iDontLikeYourChange) {
    throw 'toughLuck';
  }
  this._activationInfo = activationInfo;
}
```

[the remark this answers is not in the document]
Comment (mingxuanzhangsfdx) [reply packages/salesforcedx-utils-vscode/src/helpers/activationTracker.ts:63]: Thanks. Then why did you not use a setter function on line 55?
```
    this._activationInfo = {
      ...this._activationInfo,
      ...extensionInfo,
      activationTime
    };
```

File: packages/salesforcedx-vscode-visualforce/package.json [hunk 4]
```diff
@@ -61,7 +61,7 @@
   ],
   "scripts": {
     "bundle:extension": "npm run bundle:extension:build && npm run bundle:extension:copy",
-    "bundle:extension:build": "esbuild ./src/extension.ts  --bundle --outfile=dist/index.js --format=cjs --platform=node --external:vscode --minify",
+    "bundle:extension:build": "esbuild ./src/extension.ts  --bundle --outfile=dist/index.js --format=cjs --platform=node --external:vscode --external:@salesforce/source-tracking --external:@salesforce/core",
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-visualforce/package.json:64]: whoops...shouldn't have removed --minify.  This was a debugging step

File: packages/salesforcedx-vscode-core/src/index.ts [hunk 5]
```diff
@@ -659,7 +658,7 @@ export async function activate(extensionContext: vscode.ExtensionContext) {
     }
   };
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/index.ts:659]: This object is where you can add modules to expose in the API of the extension.

File: packages/salesforcedx-utils-vscode/src/helpers/activationTracker.ts [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+    };
+  }
+
+  async markActivationStop(activationEndDate?: Date): Promise<void> {
+    // capture date and elapsed HR time
+    const activateEndDate = activationEndDate ?? new Date();
+    const hrEnd = this.telemetryService.getEndHRTime(this._activationInfo.startActivateHrTime);
+    // getting extension info. This may take up to 10 seconds, as log record creation might be lagging from
+    // this code. All needed data have been captured for telementry, so a wait here should have no effect
+    // on quaility of telemetry data
```
Comment (CristiCanizales) [inline packages/salesforcedx-utils-vscode/src/helpers/activationTracker.ts:56]:
```suggestion
    // on quality of telemetry data
```

File: packages/salesforcedx-utils-vscode/src/helpers/activationTracker.ts [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+      extensionActivationTime: 0
+    };
+  }
+
+  async markActivationStop(activationEndDate?: Date): Promise<void> {
+    // capture date and elapsed HR time
+    const activateEndDate = activationEndDate ?? new Date();
+    const hrEnd = this.telemetryService.getEndHRTime(this._activationInfo.startActivateHrTime);
+    // getting extension info. This may take up to 10 seconds, as log record creation might be lagging from
+    // this code. All needed data have been captured for telementry, so a wait here should have no effect
```
Comment (CristiCanizales) [inline packages/salesforcedx-utils-vscode/src/helpers/activationTracker.ts:55]:
```suggestion
    // this code. All needed data have been captured for telemetry, so a wait here should have no effect
```

File: packages/salesforcedx-utils-vscode/src/helpers/activationTrackerUtils.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+export const isProcessAlive = (pid: string): boolean => {
+  try {
+    process.kill(parseInt(pid, 10), 0);
+    return true; // Process is active
+  } catch (error) {
+    return false; // Process is not active
+  }
+};
+/**
+ * given an log Uri, this function will return the exthost.log file contents
```
Comment (CristiCanizales) [inline packages/salesforcedx-utils-vscode/src/helpers/activationTrackerUtils.ts:48]:
```suggestion
 * given a log Uri, this function will return the exthost.log file contents
```

**Review by CristiCanizales — CHANGES_REQUESTED** [review]
just a couple of nits

**Comment by CristiCanizales** [comment]
QE notes:
Confirmed the data is being sent to AppInsights ✅ 
image

**Review by CristiCanizales — APPROVED** [review]
LGTM!

:shipit:

**Review by CristiCanizales — APPROVED** [review]

**Review by mingxuanzhangsfdx — APPROVED** [review]
LGTM!

**Pull request merged**